rust/src/quic/quic.rs

```diff
@@ -0,0 +1,335 @@
+/* Copyright (C) 2021 Open Information Security Foundation
+ *
+ * You can copy, redistribute or modify this Program under the terms of
+ * the GNU General Public License version 2 as published by the Free
+ * Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+use super::{
+    cyu::Cyu,
+    parser::{QuicData, QuicHeader},
+};
+use crate::applayer::{self, *};
+use crate::core::{self, AppProto, Flow, ALPROTO_FAILED, ALPROTO_UNKNOWN, IPPROTO_UDP};
+use std::ffi::CString;
+
+static mut ALPROTO_QUIC: AppProto = ALPROTO_UNKNOWN;
+
+const DEFAULT_DCID_LEN: usize = 16;
+
+#[derive(Debug)]
+pub struct QuicTransaction {
+    tx_id: u64,
+    pub header: QuicHeader,
+    pub cyu: Vec<Cyu>,
+
+    de_state: Option<*mut core::DetectEngineState>,
+    events: *mut core::AppLayerDecoderEvents,
+    tx_data: AppLayerTxData,
+}
+
+impl QuicTransaction {
+    fn new(header: QuicHeader, data: QuicData) -> Self {
+        let cyu = Cyu::generate(&header, &data.frames);
+        QuicTransaction {
+            tx_id: 0,
+            header,
+            cyu,
+            de_state: None,
+            events: std::ptr::null_mut(),
+            tx_data: AppLayerTxData::new(),
+        }
+    }
+
+    fn free(&mut self) {
+        if !self.events.is_null() {
+            core::sc_app_layer_decoder_events_free_events(&mut self.events);
+        }
+        if let Some(state) = self.de_state {
+            core::sc_detect_engine_state_free(state);
+        }
+    }
+}
+
+impl Drop for QuicTransaction {
+    fn drop(&mut self) {
+        self.free();
+    }
+}
+
+pub struct QuicState {
+    tx_id: u64,
```

Comment:
Is this not simply `transactions.len()` ?

Reply:
Oh no, never mind my comment

Reply:
we should probably rename all these fields and the API too, to indicate we're tracking the max_id